### ISSUE TITLE
Raw Sqlalchemy 

### ISSUE THREAD

**Issue by s-y** [pallets-eco/flask-debugtoolbar#71]
Hello!

I use Sqlalchemy  without flask-sqlalchemy and canot see sql-queries.

**Comment by dtheodor** [comment]
:+1:

**Comment by mgood** [comment]
So, one of the issues here seems to be primarily documentation related. The docs don't explain that you need to use the Flask-SQLAlchemy extension, and so some people are understandably confused when their queries don't show up. This could be addressed by adding more info to the docs, and also a warning in the panel itself if it detects that the Flask extension is not being used.

The other, that I'd like to understand better, is when people are aware of the Flask-SQLAlchemy extension, but would prefer not to use it. What are the issues with Flask-SQLAlchemy that are motivating you not to use it? I know that Flask-SQLAlchemy adds a couple Flask-specific things like the "Model.query" helper attribute that some people may avoid for compatibility with other non-Flask usage, although it seems like it's possible to just use the session-management aspect without the others.

I don't want to stifle use-cases that I don't yet understand, but generally it seems like most apps are simplified by using Flask-SQLAlchemy. So if it's mostly just an issue of documentation, I'd prefer to focus on that and keep both this extension, and users' apps simpler by just using the existing Flask-SQLAlchemy support.

**Comment by dtheodor** [comment]
I think you should keep using Flask-SQLAlchemy exactly as you do now (and make it a proper optional requirement for the project installable with `pip install flaskdebugtoolbar[sqlalchemy]`), but do not limit people that want to use models declared in plain SQlAlchemy.

The API of Flask-SQLAlchemy is not a superset of the SQLAlchemy API, and limits you in some ways that SQLAlchemy does not (engine configuration for example has to go through Flask app configuration, something that is a hindrance in my use-case). It's also a matter of documentation; I prefer to stick to SQL Alchemy documentation instead of having to switch between the two to figure out how something works. In general I found Flask-SQLAlchemy only a distraction and not an added value.

That's not to say that Flask-SQLAlchemy does not come with a few useful features, especially for the debugtoobar (like counting queries). This can still be used while having models declared in plain SQLAlchemy. My pull-request implements all of the above.

**Comment by mgood** [comment]
Yes, with Flask-SQLAlchemy you need to configure the engine through the Flask config. However besides that, I don't think you need to use any Flask-SQLAlchemy specific APIs if you prefer not to. I find the Flask-SQLAlchemy shortcuts useful for simplifying the code, but you could ignore them and just use the SQLAlchemy APIs and documentation directly.

Is it possible to provide an example of how you're setting up the engine that doesn't work well with Flask-SQLAlchemy? I'm trying to get a better feel for when someone else might encounter the same need to create the engine directly.

**Comment by dtheodor** [comment]
Engine is configured through Flask config, and then models declared through the `db.Model` which couples them to the engine which lies behind the `db` (unlike SQLAlchemy where models are decoupled from the engine, -1 point even here),

Nitpicking aside, this is the very simple case where I found having to use Flask-SQLAlchemy problematic: My db models and engine are in a python package independent from the Flask app (independent from anything else actually). This package is used by 1) my Flask web app, 2) my 2nd and 3rd other web apps (cubes and eve), 3) by the celery task queue, 4) by various other scheduled python scripts You can see having to couple the db/engine with my Flask app is a very counter-productive way to architect this system.

**Comment by davidism** [comment]
Over at the Flask-SQLAlchemy repo, we recognize that there are too many things coupled to specific behavior.  The *session* is actually more problematic than the *engine*, as it's what's selecting the engine in the first place.  Model shouldn't be an issue (and you can use other metadata in any given session) unless you want to use FSQLA's `__bind__` support.

The proper way for the Toolbar to support other SQLAlchemy configurations is to provide its own events that can be added to any session to record the necessary debug info.

**Comment by jmagnusson** [comment]
Has anyone made a debug panel that uses "raw" SQLalchemy? I also don't like the way that Flask-SQLAlchemy forces me to couple my db package to the web app.

**Comment by sarimak** [comment]
I also prefer raw SQLAlchemy due to the coupling (access the DB outside of Flask). I have compiled almost perfect configuration-based workaround:

from flask_sqlalchemy import _EngineDebuggingSignalEvents  # noqa
_EngineDebuggingSignalEvents(db.get_engine(app), app.import_name).register()  # Display SQL queries issued by SQLAlchemy
from flask_debugtoolbar import DebugToolbarExtension  # noqa
app.debug = True
app.config["SECRET_KEY"] = "123"
app.config["DEBUG_TB_ENABLED"] = True
app.config["DEBUG_TB_PROFILER_ENABLED"] = True
app.config["SQLALCHEMY_RECORD_QUERIES"] = True
app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = True
debug_toolbar = DebugToolbarExtension(app)  # Setup of the toolbar must be performed in advance
app.extensions["debugtoolbar"] = debug_toolbar  # Workaround for bug in flask-debugtoolbar (the extension doesn't register itself)

I can see the queries and their duration, their resultset and EXPLAIN but the Context is not correct. Maybe it is caused by https://github.com/mgood/flask-debugtoolbar/issues/96

**Comment by sarimak** [comment]
@mgood: This is how I use part of Flask-SQLAlchemy in the Flask app:

```
from flask_sqlalchemy import SQLAlchemy
db = SQLAlchemy(session_options={"autocommit": True})
db.init_app(app)

# By default, SQLAlchemy creates an implicit transaction after data change.
# Changed data then must be explicitly commited using session.commit().
# Flask-SQLAlchemy keeps the default behavior of SQLAlchemy (called autocommit=False).
# Changing data managed by SQLAlchemy inside session.begin() context with disabled autocommit would
#  create 2nd transaction and SQLAlchemy disallows nested transactions by default. But:
#  enabled nested transactions means that inner commit() is ignored until outer commit() happens.
# => Enabling autocommit is needed for using explicit transactions.

# Flask-SQLAlchemy already produces thread-safe scoped_session (app.db) but without autocommit
# Flask-SQLAlchemy couples declarative base with its Flask extension which is not desired
# See http://derrickgilland.com/posts/demystifying-flask-sqlalchemy/ for details
db_session = db.session

BaseModel = declarative_base()  # All declarative mapped classes must inherit from this instance
```

and raw SQLAlchemy in the stand-alone script:
```
engine = create_engine(connection_string)  # echo=True
session = sessionmaker(bind=self.engine)()  # No need for thread-local scoped_session
meta = MetaData()  # BaseModel.metadata would reflect imported classes, not the persisted DB
meta.reflect(self.engine)

```

**Comment by illume** [comment]
I updated the merge conflict, docs and a few things from the @dtheodor pull request into a new pull request at #110 

What do you think @jeffwidman ?
